nilearn/region_extraction/regionextractor.py

```diff
@@ -0,0 +1,232 @@
+"""
+region_extractor
+"""
+
+import nibabel
+import numpy as np
+
+from scipy.ndimage import label
+
+from skimage.feature import peak_local_max
+from skimage.segmentation import random_walker
+
+from nilearn import plotting
+from nilearn._utils import check_niimg
+from nilearn._utils.extmath import fast_abs_percentile
+from nilearn.image import iter_img, new_img_like
+from nilearn.image.image import _smooth_array
+from nilearn._utils.niimg_conversions import concat_niimgs
+from nilearn._utils.compat import _basestring
+
+
+def extract_seperate_regions(img, img_data, min_size):
+    """ Extract and seperate the regions which have largest
+        connections considering group of voxels which are
+        certainly high given a thresholded image of the
+        brain maps/components.
+
+    Parameters
+    ----------
+    img: 3D Nifti like image/object
+        contains a thresholded image of the components.
+
+    Returns
+    -------
+    regions_img: a 4D Nifti like image/object
+        contains the regions extracted from
+        the input image where each 3D volume is
+        assigned as each seperate region by
+        concatenating in the 4th dimension to a 4D image.
+    """
+    regions_img = []
+    affine = img.get_affine()
+    region = img_data.copy()
+
+    labels, n_labels = label(img_data)
+    labels_size = np.bincount(labels.ravel())
+    labels_size[0] = 0.
+
+    for label_id, label_size in enumerate(labels_size):
+        if label_size > min_size:
+            region = img_data.copy()
+            region = (labels == label_id) * img_data
+            region_img = new_img_like(img, region, affine)
+            regions_img.append(region_img)
+
+    return regions_img
+
+
+class RegionExtractor():
+    """ Region Extraction is a post processing technique which
+    is implemented to automatically extracts each brain atlas maps
+    into different set of seperated brain activated region.
+    Particularly, to show that each decomposed brain maps can be
+    used to focus on a target specific Regions of Interest analysis.
+    The idea to implement is a two step procedure:
+        1. Foreground Extraction using automatic thresholding function
+            based upon "percentile" of the features.
+            Ref: "fast_abs_percentile" in nilearn._utils.extmath
```

Comment:
You don't have to go into so much details: we are trying to hide implementation from user.